src/experimental-layers/src/solid-polygon-layer/polygon-tesselator.js

```diff
@@ -103,109 +139,112 @@ function getPolygonOffsets(polygons) {
   // Calculate length of index array (3 * number of triangles)
   const indexCount = 3 * getTriangleCount(polygons);
   const offsets = getPolygonOffsets(polygons);
 
   // Allocate the attribute
-  // TODO it's not the index count but the vertex count that must be checked
-  if (IndexType === Uint16Array && indexCount > 65535) {
-    throw new Error('Vertex count exceeds browser\'s limit');
-  }
   const attribute = new IndexType(indexCount);
 
   // 1. get triangulated indices for the internal areas
   // 2. offset them by the number of indices in previous polygons
   let i = 0;
   polygons.forEach((polygon, polygonIndex) => {
-    for (const index of calculateSurfaceIndices(polygon)) {
+    for (const index of Polygon.getSurfaceIndices(polygon)) {
       attribute[i++] = index + offsets[polygonIndex];
     }
   });
 
   return attribute;
 }
 
-/*
- * Get vertex indices for drawing complexPolygon mesh
- * @private
- * @param {[Number,Number,Number][][]} complexPolygon
- * @returns {[Number]} indices
- */
-function calculateSurfaceIndices(complexPolygon) {
-  // Prepare an array of hole indices as expected by earcut
-  const holeIndices = getHoleIndices(complexPolygon);
-  // Flatten the polygon as expected by earcut
-  const verts = flattenVertices(complexPolygon);
-  // Let earcut triangulate the polygon
-  return earcut(verts, holeIndices, 3);
-}
-
-function calculatePositions({polygons, pointCount, fp64}) {
+function updatePositions(
+  {positions, positions64xyLow, nextPositions, nextPositions64xyLow},
+  {polygons, extruded, fp64}
+) {
   // Flatten out all the vertices of all the sub subPolygons
-  const attribute = new Float32Array(pointCount * 3);
-  let attributeLow;
-  if (fp64) {
-    // We only need x, y component
-    attributeLow = new Float32Array(pointCount * 2);
-  }
   let i = 0;
   let j = 0;
-  for (const polygon of polygons) {
-    Polygon.forEachVertex(polygon, vertex => { // eslint-disable-line
+  let nextI = 0;
+  let nextJ = 0;
+  let startVertex = null;
+
+  const popStartVertex = () => {
+    if (startVertex && extruded) {
+      nextPositions[nextI++] = startVertex.x;
+      nextPositions[nextI++] = startVertex.y;
+      nextPositions[nextI++] = startVertex.z;
+      if (fp64) {
+        nextPositions64xyLow[nextJ++] = startVertex.xLow;
+        nextPositions64xyLow[nextJ++] = startVertex.yLow;
+      }
+    }
+    startVertex = null;
+  };
+
+  polygons.forEach((polygon, polygonIndex) => {
+    Polygon.forEachVertex(polygon, (vertex, vertexIndex) => { // eslint-disable-line
       const x = vertex[0];
       const y = vertex[1];
       const z = vertex[2] || 0;
-      attribute[i++] = x;
-      attribute[i++] = y;
-      attribute[i++] = z;
+      let xLow;
+      let yLow;
+
+      positions[i++] = x;
+      positions[i++] = y;
+      positions[i++] = z;
       if (fp64) {
-        attributeLow[j++] = fp64ify(x)[1];
-        attributeLow[j++] = fp64ify(y)[1];
+        xLow = x - Math.fround(x);
```

Comment:
Did removing a call to fp64ify here really make a perf difference? Readability matters so it should be justified.

In webpack named exports of small math functions does affect perf, maybe just defining `fp64ify` in this file, or exporting it as part of an object could mitigate any perf issues.